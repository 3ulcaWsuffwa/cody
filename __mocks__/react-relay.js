const ReactRelay = require("react-relay");

module.exports = {
  ...ReactRelay,
  // Mock createFragmentContainer HOC so it just returns the inner Component
  createFragmentContainer: Component => Component
};
