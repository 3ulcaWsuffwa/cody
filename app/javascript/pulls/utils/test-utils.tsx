import React from "react";
import { render } from "@testing-library/react";
import { IntlProvider } from "react-intl";
import { MemoryRouter } from "react-router";

// eslint-disable-next-line react/prop-types
const TestProvider = ({ children }) => {
  return (
    <IntlProvider locale="en" timeZone="UTC">
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </IntlProvider>
  )
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function customRender(ui, options?) {
  return render(ui, { wrapper: TestProvider, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
