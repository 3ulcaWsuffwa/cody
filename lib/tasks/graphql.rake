namespace :graphql do
  desc "Export the GraphQL schema"
  task schema: :environment do
    File.write(
      Rails.root.join("schema.graphql"),
      GraphQL::Schema::Printer.print_schema(CodySchema)
    )
  end

  desc "Run Relay compiler"
  task relay: :schema do
    sh "bin/yarn relay"
  end
end
