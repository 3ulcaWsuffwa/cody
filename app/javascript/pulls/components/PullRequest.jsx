// @flow

import React from "react";
import cn from "lib/cn";
import { createFragmentContainer, graphql } from "react-relay";
import { Link } from "react-router-dom";
import type { PullRequest_pullRequest } from "./__generated__/PullRequest_pullRequest.graphql";

const PullRequest = ({
  pullRequest: { id, number, repository, status }
}: {
  pullRequest: PullRequest_pullRequest
}) =>
  <div className={cn("-pull-request")}>
    <div>
      <div className="dib pv1 f4 code near-black lh-copy">
        {`${repository}#${number}`}
      </div>
      <div className="dib pv1 f4 code gray mh3 lh-copy">
        {status}
      </div>
    </div>
    <Link
      to={`/repos/${repository}/pull/${number}`}
      className={cn("-pull-request--more more-button")}
    >
      &bull; &bull; &bull;
    </Link>
  </div>;

export default createFragmentContainer(
  PullRequest,
  graphql`
    fragment PullRequest_pullRequest on PullRequest {
      id
      repository
      number
      status
    }
  `
);
