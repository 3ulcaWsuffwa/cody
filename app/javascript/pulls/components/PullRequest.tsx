import React from "react";
import { CheckCircle, ChevronRight, Circle, GitPullRequest, XCircle } from "react-feather";
import { createFragmentContainer, graphql } from "react-relay";
import { Link } from "react-router-dom";
import type { PullRequest_pullRequest } from "./__generated__/PullRequest_pullRequest.graphql";

function iconForStatus(status: string): JSX.Element {
  switch(status) {
    case "pending_review":
      return <Circle className="flex-shrink-0 mr-1.5 h-5 w-5 text-yellow-300" />;
    case "approved":
      return <CheckCircle className="flex-shrink-0 mr-1.5 h-5 w-5 text-green-500" />;
    case "closed":
      return <XCircle className="flex-shrink-0 mr-1.5 h-5 w-5 text-red-500" />;
  }
}

function messageForStatus(status: string): string {
  switch(status) {
    case "pending_review":
      return "Pending Review";
    case "approved":
      return "Approved";
    case "closed":
      return "Closed";
  }
}

function PullRequest({
  pullRequest: { number, repository, status },
}: {
  pullRequest: PullRequest_pullRequest;
}): JSX.Element {
  return (
    <li>
      <Link
        to={`/repos/${repository}/pull/${number}`}
        className="block hover:bg-gray-50"
      >
        <div className="flex items-center px-4 py-4 sm:px-6">
          <div className="min-w-0 flex-1 flex items-center">
            <div className="flex-shrink-0">
              <GitPullRequest />
            </div>
            <div className="min-w-0 flex-1 px-4 md:grid md:grid-cols-2 md:gap-4">
              <div>
                <p className="text-sm font-medium truncate">{`${repository}#${number}`}</p>
                <p className="mt-2 flex items-center text-sm text-gray-500" data-testid="pull-request-status">
                  {iconForStatus(status)}
                  {messageForStatus(status)}
                </p>
              </div>
            </div>
            <div>
              <ChevronRight />
            </div>
          </div>
        </div>
      </Link>
    </li>
  );
}

export default createFragmentContainer(PullRequest, {
  pullRequest: graphql`
    fragment PullRequest_pullRequest on PullRequest {
      id
      repository
      number
      status
    }
  `,
});
