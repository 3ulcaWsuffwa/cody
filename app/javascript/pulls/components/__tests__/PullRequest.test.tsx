import React from "react";
import { render, screen } from "test-utils";
import { createMockEnvironment, MockPayloadGenerator } from "relay-test-utils";
import { QueryRenderer, graphql, Environment } from "react-relay";
import PullRequest from "../PullRequest";
import "@testing-library/jest-dom/extend-expect";
import { PullRequestTestSnapshotQuery } from "./__generated__/PullRequestTestSnapshotQuery.graphql";

const TestComponent = ({ environment }: { environment: Environment }) => (
  <QueryRenderer<PullRequestTestSnapshotQuery>
    environment={environment}
    query={graphql`
      query PullRequestTestSnapshotQuery @relay_test_operation {
        viewer {
          repository(owner: "test", name: "test") {
            pullRequests(first: 10) {
              edges {
                node {
                  ...PullRequest_pullRequest
                }
              }
            }
          }
        }
      }
    `}
    variables={{}}
    render={({ error, props }) => {
      if (error) {
        return error.message;
      } else if (props) {
        return (
          <PullRequest
            pullRequest={props.viewer.repository.pullRequests.edges[0].node}
          />
        );
      } else {
        return "Loading";
      }
    }}
  />
);

let environment;
beforeEach(() => {
  environment = createMockEnvironment();
});

test("PullRequest snapshot test", () => {
  const renderedComponent = render(<TestComponent environment={environment} />);

  environment.mock.resolveMostRecentOperation((operation) =>
    MockPayloadGenerator.generate(operation)
  );

  const fragment = renderedComponent.asFragment();
  expect(fragment).toMatchSnapshot();
});

test("PullRequest generates a link to view the details", () => {
  render(<TestComponent environment={environment} />);

  environment.mock.resolveMostRecentOperation((operation) =>
    MockPayloadGenerator.generate(operation, {
      PullRequest(_, generateId) {
        return {
          id: `test-id-${generateId()}`,
          repository: "test/test",
          number: "42",
          status: "approved",
        };
      },
    })
  );

  expect(screen.queryByRole("link")).toHaveAttribute(
    "href",
    "/repos/test/test/pull/42"
  );
});

test.each([
  ["approved", "Approved"],
  ["pending_review", "Pending Review"],
  ["closed", "Closed"],
])('PullRequest with status: `%s` shows message: "%s"', (status, message) => {
  render(<TestComponent environment={environment} />);

  environment.mock.resolveMostRecentOperation((operation) =>
    MockPayloadGenerator.generate(operation, {
      PullRequest(_, generateId) {
        return {
          id: `test-id-${generateId()}`,
          repository: "test/test",
          number: "42",
          status,
        };
      },
    })
  );

  expect(screen.getByTestId("pull-request-status")).toHaveTextContent(message);
});
