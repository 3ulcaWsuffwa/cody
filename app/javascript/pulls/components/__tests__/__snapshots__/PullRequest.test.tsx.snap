// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PullRequest snapshot test 1`] = `
<DocumentFragment>
  <li>
    <a
      class="block hover:bg-gray-50"
      href="/repos/<mock-value-for-field-\\"repository\\">/pull/<mock-value-for-field-\\"number\\">"
    >
      <div
        class="flex items-center px-4 py-4 sm:px-6"
      >
        <div
          class="min-w-0 flex-1 flex items-center"
        >
          <div
            class="flex-shrink-0"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="18"
                cy="18"
                r="3"
              />
              <circle
                cx="6"
                cy="6"
                r="3"
              />
              <path
                d="M13 6h3a2 2 0 0 1 2 2v7"
              />
              <line
                x1="6"
                x2="6"
                y1="9"
                y2="21"
              />
            </svg>
          </div>
          <div
            class="min-w-0 flex-1 px-4 md:grid md:grid-cols-2 md:gap-4"
          >
            <div>
              <p
                class="text-sm font-medium truncate"
              >
                &lt;mock-value-for-field-"repository"&gt;#&lt;mock-value-for-field-"number"&gt;
              </p>
              <p
                class="mt-2 flex items-center text-sm text-gray-500"
                data-testid="pull-request-status"
              />
            </div>
          </div>
          <div>
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="9 18 15 12 9 6"
              />
            </svg>
          </div>
        </div>
      </div>
    </a>
  </li>
</DocumentFragment>
`;
